File: framework/languages/dotnet/e2e/ResilienceLifecycle/Client/Scenarios/RlB3GracefulShutdownScenario.cs
// Verifies RL-B3 Graceful Shutdown behavior.
using ResilienceLifecycle.Client.Support;
using ResilienceLifecycle.Shared;
using Zlink.HttpClient;

namespace ResilienceLifecycle.Client.Scenarios;

// RL-B3 verifies graceful shutdown avoids stale endpoint routing.
internal static class RlB3GracefulShutdownScenario
{
    public static async Task RunAsync(
        ZLinkHttpClient consumer,
        ZLinkHttpClient registry,
        ResilienceProcessManager processes,
        ZLinkHttpClient providerA,
        ZLinkHttpClient providerB)
    {
        var beforeMarker = $"rl-b3-before-{Guid.NewGuid():N}";
        var before = (await consumer.Post("/profile/request")
            .Body(new ProfileReq("fast", beforeMarker))
            .Async<ProfileRes>()).Body;
        ZlinkStreamAssert.Ensure(before.ProviderRid is "api-a" or "api-b", "RL-B3 pre-shutdown request failed.");

        await providerB.Post("/shutdown").AsyncRaw();
        for (var attempt = 0; attempt < 100; attempt++)
        {
            try
            {
                var health = await providerB.Get("/health").AsyncRaw();
                if (health.Status != 200) break;
            }
            catch
            {
                break;
            }

            await Task.Delay(100);
        }

        await processes.WaitProviderBExitedAsync();
        await registry.Post("/topology/wait")
            .Body(new TopologyWaitReq("api-b", "Ready", 0))
            .Async<TopologyEntryRes[]>();
        await ProviderTrafficProbe.WaitUntilProviderExcludedAsync(
            consumer,
            "api-b",
            "rl-b3-converge",
            "RL-B3");

        for (var i = 0; i < 12; i++)
        {
            var after = (await consumer.Post("/profile/request")
                .Body(new ProfileReq("fast", $"rl-b3-after-{i}"))
                .Async<ProfileRes>()).Body;
            ZlinkStreamAssert.Ensure(after.ProviderRid == "api-a",
                "RL-B3 request after graceful shutdown used stale api-b.");
        }

        await providerA.Post("/evidence/wait")
            .Body(new EvidenceWaitReq(["marker=rl-b3-after-"], []))
            .Async<string[]>();

        await processes.StartProviderBAsync();

        Console.WriteLine("scenario RL-B3 passed");
    }
}

File: framework/languages/node/e2e/ResilienceLifecycle/Client/Scenarios/rl-b3-graceful-shutdown-scenario.ts
// RL-B3: Graceful Shutdown 뒤 topology에서 제거한다 시나리오를 검증한다.
import type { ProfileRes } from '../../Shared/messages';
import type { ClientOptions } from '../Support/client-options';
import { postJson } from '../../../http-client';
import { startProvider } from '../Support/managed-provider';
import {
  profileReq,
  waitPeerAbsent,
  waitPeerPresent,
  waitUntilAvailable,
  waitUntilDown
} from '../Support/resilience-helpers';
import { ensure } from '../Support/scenario-assert';
import type { ScenarioState } from '../Support/scenario-state';

export async function runRlB3(options: ClientOptions, state: ScenarioState): Promise<void> {
  const before = await postJson<ProfileRes>(
    options.consumerUrl,
    '/profile/request',
    profileReq(`rl-b3-before-${Date.now()}`)
  );
  ensure(before.providerRid === 'api-a' || before.providerRid === 'api-b', 'RL-B3 pre-shutdown request failed.');

  if (state.providerBProcess !== undefined) {
    await state.providerBProcess.stop();
    state.providerBProcess = undefined;
  } else {
    await postJson(options.providerBUrl, '/shutdown');
  }
  await waitUntilDown(options.providerBUrl);
  await waitPeerAbsent(options.peerLocationUrl, 'api-b');

  for (let i = 0; i < 12; i += 1) {
    const after = await postJson<ProfileRes>(
      options.consumerUrl,
      '/profile/request',
      profileReq(`rl-b3-after-${i}`)
    );
    ensure(after.providerRid === 'api-a', 'RL-B3 request after graceful shutdown used stale api-b.');
  }
  await postJson<string[]>(options.providerAUrl, '/evidence/wait', { contains: 'marker=rl-b3-after-' });

  const restarted = startProvider({
    providerMain: options.providerMain,
    logDir: options.logDir,
    redisEndpoint: options.redisEndpoint,
    redisKeyPrefix: options.redisKeyPrefix,
    name: 'api-b-b3-restored',
    rid: 'api-b',
    httpUrl: options.providerBUrl,
    channelEndpoint: options.providerBChannelEndpoint,
    evidenceFileName: 'api-b-b3-restored.evidence.log'
  });
  await restarted.waitReady();
  state.providerBProcess = restarted;
  await waitUntilAvailable(options.providerBUrl);
  await waitPeerPresent(options.peerLocationUrl, 'api-b');

  console.log('scenario RL-B3 passed');
}
